src/umpire/strategy/ResourceAwarePool.hpp

```diff
@@ -0,0 +1,246 @@
+   * above.
+   */
+  void deallocate(void* ptr, std::size_t size) override;
+  void release() override;
+
+  std::size_t getActualSize() const noexcept override;
+  std::size_t getReleasableSize() const noexcept;
+  std::size_t getActualHighwaterMark() const noexcept;
+
+  /*!
+   * \brief Returns the current size of the pool, rounded up to alignment
+   */
+  std::size_t getAlignedSize() const noexcept;
+
+  std::size_t getAlignedHighwaterMark() const noexcept;
+
+  Platform getPlatform() noexcept override;
+  MemoryResourceTraits getTraits() const noexcept override;
+  bool tracksMemoryUse() const noexcept override;
+
+  /*!
+   * \brief Return the number of memory blocks -- both leased to application
+   * and internal free memory -- that the pool holds.
+   */
+  std::size_t getBlocksInPool() const noexcept;
+
+  /*!
+   * \brief Get the largest allocatable number of bytes from pool before
+   * the pool will grow.
+   *
+   * return The largest number of bytes that may be allocated without
+   * causing pool growth
+   */
+  std::size_t getLargestAvailableBlock() noexcept;
+
+  std::size_t getReleasableBlocks() const noexcept;
+  std::size_t getTotalBlocks() const noexcept;
+
+  void coalesce() noexcept;
+  void do_coalesce(std::size_t suggested_size) noexcept;
+
+  struct Chunk;
+
+ private:
+  void do_deallocate(Chunk* chunk, void* ptr) noexcept;
+
+  template <typename Value>
+  class pool_allocator {
+   public:
+    using value_type = Value;
+    using size_type = std::size_t;
+    using difference_type = std::ptrdiff_t;
+
+    pool_allocator() : pool{std::make_shared<util::FixedMallocPool>(sizeof(Value))}
+    {
+    }
+
+    template <typename U>
+    pool_allocator(const pool_allocator<U>& other) : pool{other.pool}
+    {
+    }
+
+    Value* allocate(std::size_t n)
+    {
+      return static_cast<Value*>(pool->allocate(n));
+    }
+
+    void deallocate(Value* data, std::size_t)
+    {
+      pool->deallocate(data);
+    }
+
+    std::shared_ptr<util::FixedMallocPool> pool;
+  };
+
+  using PointerMap = std::unordered_map<void*, Chunk*>;
+  using PendingList = std::list<Chunk*>;
+  using SizeMap =
+      std::multimap<std::size_t, Chunk*, std::less<std::size_t>, pool_allocator<std::pair<const std::size_t, Chunk*>>>;
+
+ protected:
+  /*!
+   * \brief Get the camp resource associated with a ptr
+   *
+   * \param ptr The pointer to data allocated with a ResourceAwarePool
+   */
+  Resource getResource(void* ptr) const;
+
+  /*!
+   * \brief Get the number of Pending chunks in the pool.
+   */
+  std::size_t getNumPending() const noexcept;
+
+ public:
+  struct Chunk {
+    Chunk(void* ptr, std::size_t s, std::size_t cs, Resource r) : data{ptr}, size{s}, chunk_size{cs}, m_resource{r}
+    {
+    }
+
+    void* data{nullptr};
+    std::size_t size{0};
+    std::size_t chunk_size{0};
+    bool free{true};
+    Chunk* prev{nullptr};
+    Chunk* next{nullptr};
+    SizeMap::iterator size_map_it;
+    Resource m_resource;
```

Comment:
Minor, but I would call these resource and event - m_ prefix is used for "main" class members.

Reply:
Done